src/passes/LoopInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,246 @@
+    // in a block just preceding the loop.
+    std::vector<Expression*> movedCode;
+    // Accumulate effects of things we can't move out - things
+    // we move out later must cross them, so we must verify it
+    // is ok to do so.
+    EffectAnalyzer effectsSoFar(getPassOptions());
+    // The loop's total effects also matter. For example, a store
+    // in the loop means we can't move a load outside.
+    // FIXME: we look at the loop "tail" area too, after the last
+    //        possible branch back, which can cause false positives
+    //        for bad effect interactions.
+    EffectAnalyzer loopEffects(getPassOptions(), loop);
+    // Note all the sets in each loop, and how many per index. Currently
+    // EffectAnalyzer can't do that, and we need it to know if we
+    // can move a set out of the loop (if there is another set
+    // still there, we can't). Another possible option here is for
+    // LocalGraph to track interfering sets. TODO
+    // FIXME: also the loop tail issue from above.
+    auto numLocals = getFunction()->getNumLocals();
+    std::vector<Index> numSetsForIndex(numLocals);
+    std::fill(numSetsForIndex.begin(), numSetsForIndex.end(), 0);
+    std::unordered_set<SetLocal*> loopSets;
+    {
+      FindAll<SetLocal> finder(loop);
+      for (auto* set : finder.list) {
+        numSetsForIndex[set->index]++;
+        loopSets.insert(set);
+      }
+    }
+    // Walk along the loop entrance, while all the code there
+    // is executed unconditionally. That is the code we want to
+    // move out - anything that might or might not be executed
+    // may be best left alone anyhow.
+    std::vector<Expression**> work;
+    work.push_back(&loop->body);
+    while (!work.empty()) {
+      auto** currp = work.back();
+      work.pop_back();
+      auto* curr = *currp;
+      // Look into blocks.
+      if (auto* block = curr->dynCast<Block>()) {
+        auto& list = block->list;
+        Index i = list.size();
+        if (i > 0) {
+          do {
+            i--;
+            work.push_back(&list[i]);
+          } while (i != 0);
+        }
+        continue;
+        // Note that if the block had a merge at the end, we would have seen
+        // a branch to it anyhow, so we would stop before that point anyhow.
+      }
+      // If this may branch, we are done.
+      EffectAnalyzer effects(getPassOptions(), curr);
+      if (effects.branches) {
+        break;
+      }
+      if (interestingToMove(curr)) {
+        // Let's see if we can move this out.
+        // Global side effects would prevent this - we might end up
+        // executing them just once.
+        // And we must also move across anything not moved out already,
+        // so check for issues there too.
+        // The rest of the loop's effects matter too, we must also
+        // take into account global state like interacting loads and
+        // stores.
+        if (!effects.hasGlobalSideEffects() &&
```

Comment:
I'd write this
```c++
bool unsafeToMove = effects.hasGlobalSideEffects() ||
    effectsSoFar.invalidates(effects) ||
    (effects.noticesGlobalSideEffects() && loopEffects.hasGlobalSideEffects());
if (!unsafeToMove) {
```
just to factor out the common `!`

Reply:
Good idea, fixing.